allow * to be used in registry path for remove-item

SteveL-MSFT opened the pull request:
Code is creating a DirectoryInfo instance to determine if the path is a directory or file, but it fails because `*` is not allowed, however, `*` is a valid character in a registry path.  Fix is to catch this exception and ignore it.  This means that using `-literalpath` with an asterisk and the filesystem provider will now return an error rather than returning quietly.

Fix https://github.com/PowerShell/PowerShell/issues/4704
Fix https://github.com/PowerShell/PowerShell/issues/4330

<!--

If you are a PowerShell Team member, please make sure you choose the Reviewer(s) and Assignee for your PR.
If you are not from the PowerShell Team, you can leave the fields blank and the Maintainers will choose them for you. If you are familiar with the team, feel free to mention some Reviewers yourself.

For more information about the roles of Reviewer and Assignee, refer to [CONTRIBUTING.md](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md).

-->

SteveL-MSFT committed:
[feature]
allow * to be used in registry path for remove-item

anmenaga commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3266,6 +3266,10 @@ protected override void ProcessRecord()
                     {
                         // not a directory
                     }
+                    catch (System.ArgumentException)
```
Maybe it is better to skip that <DirectoryInfo code section> altogether if provider is not FileSystem provider?
e.g.: resolvedPath.Provider.Name.Equals(FileSystemProvider.ProviderName, StringComparison.OrdinalIgnoreCase)

anmenaga replied:
... that <DirectoryInfo code section> section altogether ...

anmenaga replied:
... that "DirectoryInfo code section" section altogether ...

adityapatwardhan replied:
I agree checking for the provider seems to be a better option.

SteveL-MSFT replied:
will make the change

anmenaga commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Navigation.cs:
```diff
@@ -3266,6 +3266,10 @@ protected override void ProcessRecord()
                         // not a directory
                     }
+                    catch (System.ArgumentException)
+                    {
```
Need to make sure that FileSystem paths continue to work as they were before this new catch handler; 2 tests:
1) remove-item with -Path and filesystem path that contains *
2) remove-item with -LiteralPath and filesystem path that contains *

SteveL-MSFT replied:
Looks like there's tests for -Path, I'll add a test for -LiteralPath with *

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1:
```diff
@@ -469,6 +469,22 @@ Describe "Extended Registry Provider Tests" -Tags @("Feature", "RequireAdminOnWi
+                $testPath | Should Not Exist
+            }
+            finally {
+                Remove-Item -Recurse $tempPath
```
Please add `-ErrorAction SilentlyContinue -Force`

SteveL-MSFT replied:
will add

adityapatwardhan commented on test/powershell/Modules/Microsoft.PowerShell.Management/Registry.Tests.ps1:
```diff
@@ -469,6 +469,22 @@ Describe "Extended Registry Provider Tests" -Tags @("Feature", "RequireAdminOnWi
+            try {
+                $tempPath = "HKCU:\_tmp"
+                $testPath = "$tempPath\*\sub"
+                New-Item -Force $testPath
```
Please assign to null.
```Powershell
$null = New-Item -Force $testPath
```

SteveL-MSFT replied:
will fix

SteveL-MSFT committed:
[feature]
address PR feedback
have remove-item return error if -literalpath doesn't resolve to path

SteveL-MSFT commented:
@PowerShell/powershell-committee reviewed this and is ok with the minor `breaking change`

adityapatwardhan reviewed (approved):
LGTM

adityapatwardhan commented:
@anmenaga Can you have a look again?

adityapatwardhan commented:
@SteveL-MSFT Can you update the PR description.

SteveL-MSFT committed:
[feature]
fix in navigation exposed an issue with WSMan Config Provider tests
that require -Recurse to be used otherwise a confirmation prompt shows
up
